crates/bevy_ecs/src/storage/resource.rs

```diff
@@ -147,31 +214,42 @@ impl Resources {
 
     /// Gets read-only access to a resource, if it exists.
     #[inline]
-    pub fn get(&self, component_id: ComponentId) -> Option<&ResourceData> {
+    pub fn get(&self, component_id: ComponentId) -> Option<&ResourceData<SEND>> {
         self.resources.get(component_id)
     }
 
     /// Gets mutable access to a resource, if it exists.
     #[inline]
-    pub(crate) fn get_mut(&mut self, component_id: ComponentId) -> Option<&mut ResourceData> {
+    pub(crate) fn get_mut(&mut self, component_id: ComponentId) -> Option<&mut ResourceData<SEND>> {
         self.resources.get_mut(component_id)
     }
 
     /// Fetches or initializes a new resource and returns back it's underlying column.
     ///
     /// # Panics
     /// Will panic if `component_id` is not valid for the provided `components`
+    ///
+    /// # Safety
+    /// If `SEND` is true, `component_id` must be initialized in Components as a `Send` and
+    /// `Sync` component.
+    ///
+    /// This is checked in debug builds and will panic if violated.
     pub(crate) fn initialize_with(
         &mut self,
         component_id: ComponentId,
         components: &Components,
         f: impl FnOnce() -> ArchetypeComponentId,
-    ) -> &mut ResourceData {
+    ) -> &mut ResourceData<SEND> {
         self.resources.get_or_insert_with(component_id, || {
             let component_info = components.get_info(component_id).unwrap();
+            if SEND {
+                debug_assert!(component_info.is_send_and_sync());
```

Comment:
I feel like we probably ought to make  this a regular `assert` and this fn safe. This should be a relatively uncommon path to hit and `!Send` stuff with resources has been unsound historically, and also during this PR so reducing the ability for unsoundness might be nice.

Reply:
Made it a normal assert and had it check unconditionally to ensure both the send and non-send cases are checked.